cpp/src/arrow/array/builder_binary.h

```diff
@@ -77,56 +89,184 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
     UnsafeAppendNextOffset();
     value_data_builder_.UnsafeAppend(value, length);
     UnsafeAppendToBitmap(true);
   }
 
-  void UnsafeAppend(const char* value, int32_t length) {
+  void UnsafeAppend(const char* value, offset_type length) {
     UnsafeAppend(reinterpret_cast<const uint8_t*>(value), length);
   }
 
   void UnsafeAppend(const std::string& value) {
-    UnsafeAppend(value.c_str(), static_cast<int32_t>(value.size()));
+    UnsafeAppend(value.c_str(), static_cast<offset_type>(value.size()));
   }
 
   void UnsafeAppend(util::string_view value) {
-    UnsafeAppend(value.data(), static_cast<int32_t>(value.size()));
+    UnsafeAppend(value.data(), static_cast<offset_type>(value.size()));
   }
 
   void UnsafeAppendNull() {
     const int64_t num_bytes = value_data_builder_.length();
-    offsets_builder_.UnsafeAppend(static_cast<int32_t>(num_bytes));
+    offsets_builder_.UnsafeAppend(static_cast<offset_type>(num_bytes));
     UnsafeAppendToBitmap(false);
   }
 
-  void Reset() override;
-  Status Resize(int64_t capacity) override;
+  /// \brief Append a sequence of strings in one shot.
+  ///
+  /// \param[in] values a vector of strings
+  /// \param[in] valid_bytes an optional sequence of bytes where non-zero
+  /// indicates a valid (non-null) value
+  /// \return Status
+  Status AppendValues(const std::vector<std::string>& values,
+                      const uint8_t* valid_bytes = NULLPTR) {
+    std::size_t total_length = std::accumulate(
+        values.begin(), values.end(), 0ULL,
+        [](uint64_t sum, const std::string& str) { return sum + str.size(); });
+    ARROW_RETURN_NOT_OK(Reserve(values.size()));
+    ARROW_RETURN_NOT_OK(value_data_builder_.Reserve(total_length));
+    ARROW_RETURN_NOT_OK(offsets_builder_.Reserve(values.size()));
+
+    if (valid_bytes) {
+      for (std::size_t i = 0; i < values.size(); ++i) {
+        UnsafeAppendNextOffset();
+        if (valid_bytes[i]) {
+          value_data_builder_.UnsafeAppend(
+              reinterpret_cast<const uint8_t*>(values[i].data()), values[i].size());
+        }
+      }
+    } else {
+      for (std::size_t i = 0; i < values.size(); ++i) {
+        UnsafeAppendNextOffset();
+        value_data_builder_.UnsafeAppend(
+            reinterpret_cast<const uint8_t*>(values[i].data()), values[i].size());
+      }
+    }
+
+    UnsafeAppendToBitmap(valid_bytes, values.size());
+    return Status::OK();
+  }
+
+  /// \brief Append a sequence of nul-terminated strings in one shot.
+  ///        If one of the values is NULL, it is processed as a null
+  ///        value even if the corresponding valid_bytes entry is 1.
+  ///
+  /// \param[in] values a contiguous C array of nul-terminated char *
+  /// \param[in] length the number of values to append
+  /// \param[in] valid_bytes an optional sequence of bytes where non-zero
+  /// indicates a valid (non-null) value
+  /// \return Status
+  Status AppendValues(const char** values, int64_t length,
```

Comment:
I don't know how it was before but it seems like this method belongs mostly on the string builder specifically instead of the binary one?

Reply:
Hmm, perhaps you're right. If it were me, this method wouldn't exist at all :-)

Reply:
In any case, I'm not sure it's worth introducing another base class just to limit this method to string builders.

Reply:
SGTM.